### ISSUE TITLE
Improving OpenSim for parallel processing

### ISSUE THREAD

**Issue by jimmyDunne** [opensim-org/opensim-core#2565]
It seems like the use of OpenSim in parallel processing is becoming more common. There are things that we could improve on the OpenSim side to make it easier for people to use, for example, the Matlab parallel processing toolbox. 

There are some examples of user issues [here](https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=91&t=10895&p=30220&start=0&view=), [here](https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=91&t=8653&p=0&start=0&view=&sid=7ddd3308ba214d31aee77727ef6265a5), [here](https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=91&t=7940&p=21294&start=0&view=), and [here](https://github.com/opensim-org/opensim-core/issues/2564). 

@clnsmith, have you run into issues using OpenSim in parallel and, if so, do you have some suggestions for improvements?

**Comment by clnsmith** [comment]
All of my high throughput computing is done through the command line (linux) and I have not run into issues yet doing this with OpenSim. 

The big thing I learned from our SIMM C code that I used at UW Madison was that memory leaks and memory errors possible in C/C++ can cause new issues in parallel computing. An example, is a for loop that writes to an index one larger than the actual size of an array (This doesn't throw a error in C). This ends up writing over a random piece of memory that may only be important and cause a simulation to fail 1 out of a 1000 times. You wont catch this when running 10 simulations etc on your normal laptop. But if you run 50,000 simulations, it now becomes a problem.  

Recently some of my students here at ETH have been using the parfor command in MATLAB to run simulations in parallel on a cloud computing cluster we have here. The only issue I totally understand 
 was that import org.opensim.modeling.* doesn't seem to work within the parfor loop. So we have to use it in every call (ie model = org.opensim.modeling.Model();). I think you do have to be careful to initialize all of the parallel things within the parfor loop, so that each parallel job isn't trying to work on the same Model at the same time.

**Comment by clnsmith** [comment]
https://github.com/opensim-org/opensim-core/issues/2564

I have always found the GUI tools to be very clumsy for batch processing in MATLAB. The IK solver comment ^ makes sense. Having a version of the tools that were easy to initialize in MATLAB and alter the properties would definitely help for parallelization.

The output files are another issue for massive parallelization. 10,000+ .mot files in ASCII format actually take up a decent amount of space, and take 1-2 hours to read into MATLAB for analysis. This is especially problematic for my contact simulations where you have to save a pressure for every triange in mesh for each timestep. I have switched to hdf5 (.h5) files for simulation outputs. 
https://www.hdfgroup.org/solutions/hdf5/
They have a hierarchical folder structure and save data in binary, so they are smaller and much faster to read. I have C++ code that writes data to .h5 files that I could share. Would take some work to integrate into .osim source as you need to build the hdf5 libraries to link against.
